### ISSUE TITLE
Include FFmpeg in exported application?

### ISSUE THREAD

**Issue by philipp-lehmann** [hamoid/video_export_processing#67]
Hey there

Thanks for the effort putting together this awesome library. 
I ran into troubles when exporting a standalone application from processing.

Can you include FFmpeg into the export, so that you don't have to install it separately when distributing the app?
I see that you can change the location, would it be possible to install ffmpeg within the data folder and point to this location?
Do you have other ideas?

Best Philipp

**Comment by hamoid** [comment]
Oh good point! Gee... I never considered exporting the application.

The library looks for its config file inside the Processing libraries folder, under `VideoExport/settings.json`
That file contains the path to ffmpeg. But I have no idea currently what happens with an exported app. I guess it still searches for that same file? Can you look if there is such a file in your exported app which you could edit?

**Comment by hamoid** [comment]
I see that after running the exported application once, a settings file is created in the same folder. You could try to put ffmpeg in a subfolder and then edit that settings.json file to point to it. A relative path might work.

**Comment by philipp-lehmann** [comment]
Yes, I had the same idea, but couldn't fully test it yet. 
Here's what i have observed so far with exported applications:

**Win 10** 
The dialog pop up and asks for the file to be located, and then runs correctly. 

**Mac OS**
If FFmpeg is installed in the default directory, everything works as expected. 
If not the application crashes.

I'll let you know if the relative paths work.

**Comment by hamoid** [comment]
I hope that by setting the path in settings.json the Environmental Variables are not needed.

Yes let me know how it goes. If it does I could add it to the readme :)

**Comment by philipp-lehmann** [comment]
Hey there

Here are my current findings

**Windows**
Works with relative paths.

**MacOS** 
Works but you have to do some trickery. Right click on the app and show the package content. 
Move your binary files and the settings files into the MacOS directory.

<img width="1355" alt="Bildschirmfoto 2021-05-04 um 15 00 54" src="https://user-images.githubusercontent.com/4519837/117007453-facc9100-ace9-11eb-81ed-c91e603b2f6a.png">
<img width="1355" alt="Bildschirmfoto 2021-05-04 um 15 01 04" src="https://user-images.githubusercontent.com/4519837/117007458-fd2eeb00-ace9-11eb-8a42-e18b7513b924.png">

**Comment by philipp-lehmann** [comment]
**MacOS**

I think i have found a way. You can include the binaries in the exported app. 
But the path in the settings.json has to be absolute.

Since this changes when you move the App, I update the settings file when launching the App.
I made a Settings.pde file with these functions. 

`void updateSettings() {
	String AppName = "App";
	String ffmpegPath = sketchPath(AppName + ".app/Contents/bin/ffmpeg");
	JSONObject json = loadJSONObject("settings.json");
	json.setString("ffmpeg_path", ffmpegPath);
	saveJSONObject(json, AppName + ".app/Contents/settings.json");
}

public String sketchPath(String where) {
	if (sketchPath() == null) {
		return where;
	}
	try {
		if (new File(where).isAbsolute()) return where;
	} catch (Exception e) {}
	return sketchPath() + File.separator + where;
}`


<img width="1018" alt="Bildschirmfoto 2021-05-11 um 15 24 58" src="https://user-images.githubusercontent.com/4519837/117823184-7b047080-b26d-11eb-8dbd-4db771ef275e.png">

**Comment by philipp-lehmann** [comment]
And when testing the sketch it creates a folder 'App.app/Contents' - not so nice.

**Comment by gritschu** [comment]
I am happy to find that I am not the only person who needs this. @philipp-lehmann thanks for your solution. However this doesn't seem to work for me. I am not sure why yet. The creation of the absolute path seems to work fine, but I think for me the call `loadJSONObject("settings.json")` seems to fail, thus I am not able to change the content of the json file via code at the moment. The folder structure is the same as in your screenshot if I don't miss something. Does someone have any idea? Thank you!

**Comment by philipp-lehmann** [comment]
Hey there

Clarifying on my earlier answer: There are two settings.json files. One is in the Data folder. On start I get the contents of this file as JSON, update the current app path, and the save it directly as a separate file into the App Folder.

```
void updateSettings() {
	updateAppPath();
	updateTargetPath();
}

// App path and ffmpeg location
void updateAppPath() {
	JSONObject ffmpgJSON = loadJSONObject("settings.json");
	String ffmpegPath = sketchPath(AppName + ".app/Contents/bin/ffmpeg");
	ffmpgJSON.setString("ffmpeg_path", ffmpegPath);
	saveJSONObject(ffmpgJSON, AppName + ".app/Contents/settings.json");
}

public String sketchPath(String where) {
	if (sketchPath() == null) {
		return where;
	}
	try {
		if (new File(where).isAbsolute()) return where;
	} catch (Exception e) {}
	return sketchPath() + File.separator + where;
}

// Target folder
void updateTargetPath() {
	JSONObject targetJSON = loadJSONObject("target.json");
	println("Target: " + targetJSON.getString("target_folder"));
	targetFolder = targetJSON.getString("target_folder");
}

void setTargetFolder() {
	selectFolder("Select a folder to write to:", "folderSelected");
}

void folderSelected(File selection) {
	if (selection == null) {
		println("No folder selected...");
	} else {
		// Apply target folder variable
		targetFolder = selection.getAbsolutePath() + "/";
		println("Target: " + targetFolder);

		// Create and save new target json
		JSONObject targetJSON = new JSONObject();
		targetJSON.setString("target_folder", targetFolder);
		targetJSON.setString("Test", "testvalue");
		saveJSONObject(targetJSON,  AppName + ".app/Contents/Java/data/target.json");
	}
}

```

**Comment by philipp-lehmann** [comment]
I remember I also had the Issue, that the dialogue was hidden. And as you can see my current version is different, so I probably adressed this.

**Comment by gritschu** [comment]
@philipp-lehmann thank you so much!
I understand everything concerning the settings.json now.

However, I don't see a crucial difference in your current version from the one before which shall prevent the reappearing popup asking for selection of the ffmpeg path even though it has been updated inside the settings.json file.

The additional parts about the target path using the target.json has nothing to do with the issue, or does it?
If so, I don't understand where the original target.json file is coming from.

**Comment by gritschu** [comment]
Further observation:
If I change the path in the json file manually (open in editor, change, save) I don't have the problem.
I also don't have the problem, when I change the path in the json via code, then close my app, then reopen and export the movie.
The popup doesn't appear anymore and everything is fine.
Only if I change the path via
`saveJSONObject(ffmpgJSON, AppName + ".app/Contents/settings.json");`
and then try to export immediately after the popup opens again (which I want to prevent).

**Comment by philipp-lehmann** [comment]
Yes, the target.json is a directory where I export the movies to. 
This is not related to the ffmpeg import.
